### ISSUE TITLE
journalctl --until is ignored when --after-cursor and --lines are both specified

### ISSUE THREAD

**Issue by nrwahl2** [systemd/systemd#31776]
### systemd version the issue has been seen with

systemd-254.10-1.fc39.x86_64

### Used distribution

Fedora 39

### Linux kernel version used

6.6.13-200.fc39.x86_64

### CPU architectures issue was seen on

x86_64

### Component

journalctl

### Expected behaviour you didn't see

When `--after-cursor`, `--until`, and `--lines` are specified, the output should not contain any messages from after the `--until` timestamp. In particular, if `--after-cursor` comes from a time after the `--until` timestamp, the output should be empty.

```
$ sudo journalctl --after-cursor='s=f8b30a9d1e664c7d8e29a1194ee14a60;i=18bee8c;b=920817753cd14eb2b5a1efa648ddd95e;m=2aabbe63dbe;t=6139a47fa1895;x=d700d213de4d236d' --until '2024-03-14 01:00' --lines 5
-- No entries --
```

### Unexpected behaviour you saw

When `--after-cursor`, `--until`, and `--lines` are specified, `--until` is ignored. Lines from after the `--until` timestamp are shown.

```
$ sudo journalctl --after-cursor='s=f8b30a9d1e664c7d8e29a1194ee14a60;i=18bee8c;b=920817753cd14eb2b5a1efa648ddd95e;m=2aabbe63dbe;t=6139a47fa1895;x=d700d213de4d236d' --until '2024-03-14 01:00' --lines 5
Mar 14 01:01:01 laptop CROND[1549786]: (root) CMD (run-parts /etc/cron.hourly)
Mar 14 01:01:01 laptop run-parts[1549789]: (/etc/cron.hourly) starting 0anacron
Mar 14 01:01:01 laptop anacron[1549802]: Anacron started on 2024-03-14
Mar 14 01:01:01 laptop anacron[1549802]: Normal exit (0 jobs run)
Mar 14 01:01:01 laptop run-parts[1549804]: (/etc/cron.hourly) finished 0anacron
```

This is a regression.
Known working version: systemd-252-18.el9.x86_64 (RHEL 9.3)

### Steps to reproduce the problem

1. Find the cursor for some timestamp X. In this example, I use `'2024-03-14 01:01'`.
```
$ sudo journalctl --until '2024-03-14 01:01' --lines 0 --show-cursor
-- cursor: s=f8b30a9d1e664c7d8e29a1194ee14a60;i=18bee8c;b=920817753cd14eb2b5a1efa648ddd95e;m=2aabbe63dbe;t=6139a47fa1895;x=d700d213de4d236d
```

2. Query the logs for messages after the cursor and until (X - 1min) -- here, `'2024-03-14 01:00'`. Also set the `--lines` argument to some value (here, `5`).
```
$ sudo journalctl --after-cursor='s=f8b30a9d1e664c7d8e29a1194ee14a60;i=18bee8c;b=920817753cd14eb2b5a1efa648ddd95e;m=2aabbe63dbe;t=6139a47fa1895;x=d700d213de4d236d' --until '2024-03-14 01:00' --lines 5
```

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Additional info:

It works correctly when --lines is omitted (no output, because the --after-cursor comes after the --until timestamp):

$ sudo journalctl --after-cursor='s=f8b30a9d1e664c7d8e29a1194ee14a60;i=18bee8c;b=920817753cd14eb2b5a1efa648ddd95e;m=2aabbe63dbe;t=6139a47fa1895;x=d700d213de4d236d' --until '2024-03-14 01:00'
-- No entries --


It also works correctly when --after-cursor is omitted (most recent 5 lines before --until):

$ sudo journalctl --until '2024-03-14 01:00' --lines 5
Mar 14 00:59:03 laptop slack.desktop[4178081]: [03/14/24, 00:59:03:685] info: [DND_V2] (E030G10V24F) Will check for changes in DND status again in 5 minutes
Mar 14 00:59:40 laptop slack.desktop[4178081]: [03/14/24, 00:59:40:629] info: [RTM] Setting a timeout of 53 ms to process more rtm events
Mar 14 00:59:40 laptop slack.desktop[4178081]: [03/14/24, 00:59:40:682] info: [RTM] Waited 53 ms, processing more rtm events now
Mar 14 00:59:52 laptop slack.desktop[4178081]: [03/14/24, 00:59:52:464] info: [RTM] Setting a timeout of 78 ms to process more rtm events
Mar 14 00:59:52 laptop slack.desktop[4178081]: [03/14/24, 00:59:52:543] info: [RTM] Waited 78 ms, processing more rtm events now


Or when --until is omitted (first 5 lines after the --after-cursor). This is the same as the buggy behavior when --until is specified, but it's correct in this context:

$ sudo journalctl --after-cursor='s=f8b30a9d1e664c7d8e29a1194ee14a60;i=18bee8c;b=920817753cd14eb2b5a1efa648ddd95e;m=2aabbe63dbe;t=6139a47fa1895;x=d700d213de4d236d' --lines 5
Mar 14 01:01:01 laptop CROND[1549786]: (root) CMD (run-parts /etc/cron.hourly)
Mar 14 01:01:01 laptop run-parts[1549789]: (/etc/cron.hourly) starting 0anacron
Mar 14 01:01:01 laptop anacron[1549802]: Anacron started on 2024-03-14
Mar 14 01:01:01 laptop anacron[1549802]: Normal exit (0 jobs run)
Mar 14 01:01:01 laptop run-parts[1549804]: (/etc/cron.hourly) finished 0anacron
```

**Comment by nrwahl2** [comment]
A fix might be to change from this
```
if (arg_until_set && !arg_reverse && (arg_lines < 0 || arg_since_set)) {
```

to this
```
if (arg_until_set && !arg_reverse && (arg_lines < 0 || arg_since_set || arg_cursor || arg_after_cursor || arg_cursor_file)) {
```

here:
https://github.com/systemd/systemd/blob/f4a63ce2/src/journal/journalctl.c#L2102


It may be useful to have a global `use_cursor` boolean or something. Checking all three `arg` variables in `show()` isn't ideal since we already check all three in `run()`.

**Comment by nrwahl2** [comment]
> Maybe prep a draft PR in reviewable form if you're going to propose a source level fix?

I haven't had time to build and test, just tossing out a quick idea. Feel free to ignore.

**Comment by yuwata** [comment]
Cannot reproduce with 253.15-2.fc38.

**Comment by nrwahl2** [comment]
May have been introduced by 81fb5375

**Comment by yuwata** [comment]
> May have been introduced by [81fb537](https://github.com/systemd/systemd/commit/81fb5375b3b3bfc22d023d7908ad9eee4b3c1ffb)

Yeah, I think so. The commit does not take into account about --cursor or friends.

Anyway, I am working on the issue, will submit a fix with test soon, hopefully.

**Comment by yuwata** [comment]
Fix is waiting in #31861.
